mining/src/mempool/revalidate_high_priority_transactions.rs

```diff
@@ -0,0 +1,53 @@
+use crate::{
+    mempool::{
+        errors::{RuleError, RuleResult},
+        model::pool::Pool,
+        Mempool,
+    },
+    model::topological_index::TopologicalIndex,
+};
+use consensus_core::tx::{MutableTransaction, TransactionId};
+use kaspa_core::debug;
+
+impl Mempool {
+    pub(crate) fn revalidate_high_priority_transactions(&mut self) -> RuleResult<Vec<TransactionId>> {
+        // First establish a topologically ordered list of all high priority transaction ids
+
+        // Processing the transactions in a parent to chained order guarantees that
+        // any transaction removal will propagate to all chained dependencies saving
+        // validations calls to consensus.
+        let ids = self.transaction_pool.index(true).topological_index()?;
+        let mut valid_ids = vec![];
+
+        for transaction_id in ids.iter() {
+            // Try to take the transaction out of the storage map so we can mutate it with some self functions.
+            // The redeemers of removed transactions are removed too so the following call may return a None.
+            if let Some(mut transaction) = self.transaction_pool.all_mut().remove(transaction_id) {
+                let is_valid = self.revalidate_transaction(&mut transaction.mtx)?;
+                // Then put the transaction back into the storage map.
+                self.transaction_pool.all_mut().insert(*transaction_id, transaction);
+                if is_valid {
+                    // A following transaction should not remove this one from the pool since we process
+                    // in topological order
+                    // TODO: consider the scenario of two high priority txs sandwiching a low one
```

Comment:
Explain that in this case topology order is not guaranteed since we topologically sorted only high-prio